internal/controller/iprange_controller.go

```diff
@@ -0,0 +1,338 @@
+const LastIpRangeMetadataAnnotationName = "iprange.netbox.dev/last-ip-range-metadata"
+
+// IpRangeReconciler reconciles a IpRange object
+type IpRangeReconciler struct {
+	client.Client
+	Scheme            *runtime.Scheme
+	NetboxClient      *api.NetboxClient
+	Recorder          record.EventRecorder
+	OperatorNamespace string
+	RestConfig        *rest.Config
+}
+
+//+kubebuilder:rbac:groups=netbox.dev,resources=ipranges,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=netbox.dev,resources=ipranges/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=netbox.dev,resources=ipranges/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *IpRangeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	debugLogger := logger.V(4)
+
+	logger.Info("reconcile loop started")
+
+	o := &netboxv1.IpRange{}
+	err := r.Client.Get(ctx, req.NamespacedName, o)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// if being deleted
+	if !o.ObjectMeta.DeletionTimestamp.IsZero() {
+		if o.Spec.PreserveInNetbox {
+			// if there's a finalizer remove it and return
+			// this can be the case if a CR used to have spec.preserveInNetbox set to false
+			return ctrl.Result{}, r.removeFinalizer(ctx, o)
+		}
+
+		return r.deleteFromNetboxAndRemoveFinalizer(ctx, o)
+	}
+
+	// if PreserveIpInNetbox flag is false then register finalizer if not yet registered
+	if !o.Spec.PreserveInNetbox && !controllerutil.ContainsFinalizer(o, IpRangeFinalizerName) {
+		debugLogger.Info("adding the finalizer")
+		controllerutil.AddFinalizer(o, IpRangeFinalizerName)
+		if err := r.Update(ctx, o); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// 1. try to lock lease of parent prefix if IpRange status condition is not true
+	// and IpRange is owned by an IpRangeClaim
+	or := o.ObjectMeta.OwnerReferences
+	var ll *leaselocker.LeaseLocker
+	if len(or) > 0 /* len(nil array) = 0 */ && !apismeta.IsStatusConditionTrue(o.Status.Conditions, "Ready") {
+		res, err := r.tryLockOnParentPrefix(ctx, ll, o)
+		if err != nil {
+			return res, err
+		}
+	}
+
+	// 2. reserve or update ip range in netbox
+	accessor := apismeta.NewAccessor()
+	annotations, err := accessor.Annotations(o)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	ipRangeModel, err := r.generateNetboxIpRangeModelFromIpRangeSpec(ctx, o, req, annotations[LastIpRangeMetadataAnnotationName])
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	netboxIpRangeModel, err := r.NetboxClient.ReserveOrUpdateIpRange(ipRangeModel)
+	if err != nil {
+		err := r.SetConditionAndCreateEvent(ctx, o, netboxv1.ConditionIpRangeReadyFalse,
+			corev1.EventTypeWarning, fmt.Sprintf("%s-%s", o.Spec.StartAddress, o.Spec.EndAddress), err)
+		if err != nil {
+			return ctrl.Result{}, nil
```

Comment:
Quick question, is there a reason that we are ignoring the error here? :)

Reply:
Thanks for leaving a note! I overlooked this one, changed it to return the error if the status update fails.

Reply:
:)